### PR TITLE
do not watch noobaa CRs if SKIP_NOOBAA_CRD_WATCH is set to true

### PR DESCRIPTION
**Pull request by iamniting**
If the CRD is not present, it can prevent the ocs-operator from
starting which can cause problems for MS. Since MS is not going
to install the noobaa CSV anymore, it is better to only watch the
noobaa CRs if the SKIP_NOOBAA_CRD_WATCH is set to false.

MS will set SKIP_NOOBAA_CRD_WATCH env in the sub which will be
populated to the ocs-operator via OLM.

Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=2182041

### CODE REVIEW

File: controllers/storagecluster/reconcile.go [hunk 1]
```diff
@@ -281,6 +288,44 @@ func (r *StorageClusterReconciler) validateStorageClusterSpec(instance *ocsv1.St
...
 
+func (r *StorageClusterReconciler) doesCrdExist(crdName string) (bool, error) {
+	crd := extensionsv1.CustomResourceDefinition{}
+	err := r.Client.Get(r.ctx, types.NamespacedName{Name: crdName}, &crd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
```
Comment (bindrad) [inline controllers/storagecluster/reconcile.go:299]:
```suggestion
	if err := r.Client.Get(r.ctx, types.NamespacedName{Name: crdName}, &crd); err != nil {
		if errors.IsNotFound(err) {
			return false, nil
		}
		return false, err
	}
```

File: controllers/storagecluster/reconcile.go [hunk 2]
```diff
@@ -281,6 +288,44 @@ func (r *StorageClusterReconciler) validateStorageClusterSpec(instance *ocsv1.St
...
+		// crd was not available while starting lets check it
+		exists, err := r.doesCrdExist(crdName)
+		if err != nil {
+			return err
+		}
+		// crd is created restart the operator and add it to watch list
+		if exists {
+			r.Log.Info("New CRD is created, Exiting to getting it added to the watchlist.")
+			os.Exit(1)
+		}
```
Comment (bindrad) [inline controllers/storagecluster/reconcile.go:323]:
```suggestion
		if exists, err := r.doesCrdExist(crdName); err != nil {
			return err
		} else if exists {
			// crd is created restart the operator and add it to watch list
			r.Log.Info("New CRD is created, Exiting to getting it added to the watchlist.")
			os.Exit(1)
		}
```

**Review by bindrad — CHANGES_REQUESTED** [review]

**Comment by bindrad** [comment]
@iamniting Did you check if any other CRDs are not created as part of OCS Operator bundle but OCS Operator watches them?

**Comment by iamniting** [comment]
> @iamniting Did you check if any other CRDs are not created as part of OCS Operator bundle but OCS Operator watches them?

No other CR is coming from noobaa which we are watching. So we are good.

**Review by bindrad — APPROVED** [review]

**Review by umangachapagain — CHANGES_REQUESTED** [review]
I do not like this approach of restarting operator when CRDs are added. If operator requires a CRD, it should be created during deployment. If you want to add/remove watches dynamically, controller has a feature for that. Gating normal operator functions for a very specific use case isn't acceptable.

/do-not-merge

**Comment by nb-ohad** [comment]
@umangachapagain 
> ...controller has a feature for that

FYI, you can call `watch` on a controller even after the start is called, but these new watches will not populate the cache. We have a similar (and even bigger) problem with the new MS agent where offerings are selected dynamically through the lifetime of the operator and could not find any other valid solution to the problem. If there is any we would love to learn about one and I believe Nithin would love that as well. 

The reason that a restart is needed here is that ODF operator has no guarantees to install the noobaa operator before ocs-operator. Is that is solved somehow then there will be no need for a restart.

**Comment by umangachapagain** [comment]
> @umangachapagain
> 
> > ...controller has a feature for that
> 
> FYI, you can call `watch` on a controller even after the start is called, but these new watches will not populate the cache. We have a similar (and even bigger) problem with the new MS agent where offerings are selected dynamically through the lifetime of the operator and could not find any other valid solution to the problem. If there is any we would love to learn about one and I believe Nithin would love that as well.
> 
>
Looks like a bug or potential feature in upstream controller-manager. Are they recommending restarting operators as an acceptable alternative? What other options did we try? I'd be happy to take a look at your test results and provide my comments. There's a 99.99% chance that I am wrong/unreasonable, I'd like to take this opportunity to learn as well.

> The reason that a restart is needed here is that ODF operator has no guarantees to install the noobaa operator before ocs-operator. Is that is solved somehow then there will be no need for a restart.
>
Looks like a request for OLM team. I am sure they have a way to guarantee that dependencies are installed before the operator that needs those. Or atleast a guarantee that operator can start even if the dependencies are not fulfilled.

**Comment by nb-ohad** [comment]
> Looks like a bug or potential feature in upstream controller-manager. 
I agree this should be a potential feature. The way operator SDK is structured right now, a reconciler doesn't even have a way to access the controller that runs it. That design decision was made on purpose (according to docs) to prevent any reconciliation code to cancel the context and as a result "kill" the controller. 

We tried the following
- Capture the controller and push it into the reconciler then call watch - watches did not populate the cache
- Capture the controller and push it into the reconciler then cancel the controller context (not request context) - controller killed but undefined behavior for existing watches.
- Capture the context of the manager and cancel it with the intent to create another manager - many issues.
- kill the pod - the only thing that reliably worked.

We might be able to resolve some of the issues around the non-kill pod options but as of time constraints and the need for a workaround, we went with this approach for now.

> Looks like a request for OLM team.
I agree, but, according to my understanding, the OLM team is not indenting to add new features until OLM 2.0 gets released. The issue here is that ODF operator depends on both noobaa operator and ocs operator, and the current OLM feature set does not allow defining an order relationship between these 2 dependencies. Add to that the fact that the odf-operator is manually managing dependent subscriptions which also complicates things even further as this mechanism does not support an order relationship as well.

File: controllers/storagecluster/storagecluster_controller.go [hunk 3]
```diff
@@ -91,6 +91,10 @@ type StorageClusterReconciler struct {
 	IsNoobaaStandalone bool
 }
 
+func skipNoobaaCrdWatch() bool {
+	return os.Getenv("SKIP_NOOBAA_CRD_WATCH") == "true"
```
Comment (sp98) [inline controllers/storagecluster/storagecluster_controller.go:95]: where is this `SKIP_NOOBAA_CRD_WATCH`  coming from?

Comment (sp98) [reply controllers/storagecluster/storagecluster_controller.go:95]: Can it be more generic instead of being specific to Nooba?

Comment (umangachapagain) [reply controllers/storagecluster/storagecluster_controller.go:95]: It'd need to be a list of CRDs to allow that. Any generic name would just cause confusion.
We could have another variable for OB/OBC?

File: controllers/storagecluster/storagecluster_controller.go [hunk 4]
```diff
@@ -91,6 +91,10 @@ type StorageClusterReconciler struct {
 	IsNoobaaStandalone bool
 }
 
+func skipNoobaaCrdWatch() bool {
```
Comment (umangachapagain) [inline controllers/storagecluster/storagecluster_controller.go:94]: No need to make this a function, just use it inline.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/storagecluster_controller.go (hunk 3 above)
Comment (iamniting) [reply controllers/storagecluster/storagecluster_controller.go:95]: +1 to @umangachapagain 

This variable will be coming from the sub. Ms will add this variable in the sub and OLM will inject the variable into the operators.

File: controllers/storagecluster/storagecluster_controller.go (hunk 4 above)
Comment (iamniting) [reply controllers/storagecluster/storagecluster_controller.go]: removed it

**Review by bindrad — APPROVED** [review]

File: controllers/storagecluster/storagecluster_controller.go (hunk 3 above)
Comment (sp98) [reply controllers/storagecluster/storagecluster_controller.go]: ok. @umangachapagain Will that be a different PR for OB/OBC?

Comment (iamniting) [reply controllers/storagecluster/storagecluster_controller.go]: It depends on what we decide on. Mainly it depends on rook how it wants to proceed with the fix.

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**